pex/environment.py

```diff
@@ -269,16 +273,45 @@ def path(self):
         # type: () -> str
         return self._pex
 
-    def iter_distributions(self):
-        # type: () -> Iterator[FingerprintedDistribution]
-        internal_cache = os.path.join(self._pex, self._pex_info.internal_cache)
-        with TRACER.timed("Searching dependency cache: %s" % internal_cache, V=2):
-            for distribution_name, fingerprint in self._pex_info.distributions.items():
-                dist_path = os.path.join(internal_cache, distribution_name)
-                yield FingerprintedDistribution(
-                    distribution=Distribution.load(dist_path),
-                    fingerprint=fingerprint,
+    @property
+    def source_pex(self):
+        # type: () -> str
+        return self._source_pex or self._pex
+
+    def iter_distributions(self, result_type_wheel_file=False):
+        # type: (bool) -> Iterator[FingerprintedDistribution]
+        if result_type_wheel_file:
+            if not self._pex_info.deps_are_wheel_files:
+                raise ResolveError(
+                    "Cannot resolve .whl files from PEX at {pex}; its dependencies are in the "
+                    "form of pre-installed wheel chroots.".format(pex=self.source_pex)
                 )
```

Comment:
In general I only like to add orthogonal features that can be combined freely. There are 3 Pex resolver sources: Pip, lock files and PEX repositories, and 2 PEX dependency formats: installed wheel chroots and (now) raw `.whl` files. In this 6 element matrix, only this case of resolving raw `.whl` files from a PEX repository containing dependencies in installed wheel chroot form cannot be satisfied currently (this is tested explicitly in the changes to `test_pex_repository_resolve.py`).

The `pep_427.py` work that now powers `.whl` file installs instead of `pip install` does make it possible to plug this gap in a later release though with some modification of `pep_376.py`. This will also allow the existing `PEX_TOOLS=1 ./a.pex repository extract ...` tool to produce fully correct wheels for the installed wheel chroot dependencies in PEXes of that style.